### DISCUSSION TITLE
Mason Package for fish-lsp

### DISCUSSION THREAD

**Discussion by bnwa** [ndonfris/fish-lsp#55] [Show and tell]
Have created an unofficial [Mason registry](https://github.com/bnwa/mason-registry) for installing `fish-lsp` as a Mason package. Mason is an external binary package manager for [Neovim](https://github.com/neovim/neovim) whose chief virtue is permitting search, installation, and updating for tooling such as LSP servers and automatically and exclusively adding them to the runtime path for vim and neovim.

Wasn't sure if this was OK with the team here, but if so, could continue with PRs to the official Mason registry. I also notice this fish-lsp is not known to the `nvim-lspconfig` package popular with more casual Neovim users for quick and easy LSP setup, so that would be another hole to fill.

**Comment by ndonfris** [comment]
Hi @bnwa!

That's awesome news! Definitely feel free to keep submitting PRs to the mason registry! Any form of improvement towards easing the installation of this project is greatly appreciated. If there's anything you have questions on in the future, I'll do my best to get back to you and offer some help.

On the note about `nvim-lspconfig`, I believe I recently saw someone [submitted a PR](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#fish_lsp) into the project, that describes the fish-lsp's usage under the [server_configurations](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md) documentation. Is there another repo you were mentioning?

**Reply by bnwa** [reply]
> On the note about `nvim-lspconfig`, I believe I recently saw someone [submitted a PR](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md#fish_lsp) into the project, that describes the fish-lsp's usage under the [server_configurations](https://github.com/neovim/nvim-lspconfig/blob/master/doc/server_configurations.md) documentation. Is there another repo you were mentioning?

Good catch, I must've missed it when I'd scanned the repo configs. Will update on my end.

**Reply by OriLiMu** [reply]
how to install it in neovim I cant find in mason.

**Reply by ndonfris** [reply]
There is generally two options for installing the `fish-lsp` for your use case:

1. Use the [unofficial mason registry](https://github.com/bnwa/mason-registry), which @bnwa maintains. 
    > NOTE:  
    > I do not use mason.nvim, so I **can not** confirm if this method is currently working.  

2. Install with a [package manager](https://github.com/ndonfris/fish-lsp#use-a-package-manager), and use the [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig/blob/master/doc/configs.md#fish_lsp) guide to setup the
   language server.

Either method should be relatively straight forward, but if you run into further
issues feel free to [open a new issue](https://github.com/ndonfris/fish-lsp/issues)

**Reply by OriLiMu** [reply]
> There is generally two options for installing the `fish-lsp` for your use case:
> 
>     1. Use the [unofficial mason registry](https://github.com/bnwa/mason-registry), which @bnwa maintains.
>        > NOTE:
>        > I do not use mason.nvim, so I **can not** confirm if this method is currently working.
> 
>     2. Install with a [package manager](https://github.com/ndonfris/fish-lsp#use-a-package-manager), and use the [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig/blob/master/doc/configs.md#fish_lsp) guide to setup the
>        language server.
> 
> 
> Either method should be relatively straight forward, but if you run into further issues feel free to [open a new issue](https://github.com/ndonfris/fish-lsp/issues)

Neither work for me. 
For the method 2: 
I get error msg: 
![图片](https://github.com/user-attachments/assets/a0c4b0d6-d1e7-402d-97f3-f8536cf8cf73)

**LspInfo** output: said no clients attached to the buffer. I go to fish config file. and open `config.fish`
```
lspconfig: require("lspconfig.health").check()

LSP configs active in this session (globally) ~
- Configured servers: clangd, lua_ls, fish_lsp, tailwindcss, basedpyright
- OK Deprecated servers: (none)

LSP configs active in this buffer (bufnr: 1) ~
- Language client log: ~/.local/state/nvim/lsp.log
- Detected filetype: `fish`
- 0 client(s) attached to this buffer
- Other clients that match the "fish" filetype:
- Config: fish_lsp
  filetypes:         fish
  cmd:               ~/.local/share/nvim/mason/bin/fish-lsp start
  version:           `?` (Failed to get version) Tried:
  `/home/lizhe/.local/share/nvim/mason/bin/fish-lsp --version`
  `/home/lizhe/.local/share/nvim/mason/bin/fish-lsp -version`
  `/home/lizhe/.local/share/nvim/mason/bin/fish-lsp version`
  `/home/lizhe/.local/share/nvim/mason/bin/fish-lsp --help`
  
  executable:        true
  autostart:         true
  root directory:    ~/.config/

Docs for active configs: ~
- fish_lsp docs: >markdown
  
  https://github.com/ndonfris/fish-lsp
  
  A Language Server Protocol (LSP) tailored for the fish shell.
  This project aims to enhance the coding experience for fish,
  by introducing a suite of intelligent features like auto-completion,
  scope aware symbol analysis, per-token hover generation, and many others.
  
  [homepage](https://www.fish-lsp.dev/)
```

**Reply by OriLiMu** [reply]
> There is generally two options for installing the `fish-lsp` for your use case:
> 
>     1. Use the [unofficial mason registry](https://github.com/bnwa/mason-registry), which @bnwa maintains.
>        > NOTE:
>        > I do not use mason.nvim, so I **can not** confirm if this method is currently working.
> 
>     2. Install with a [package manager](https://github.com/ndonfris/fish-lsp#use-a-package-manager), and use the [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig/blob/master/doc/configs.md#fish_lsp) guide to setup the
>        language server.
> 
> 
> Either method should be relatively straight forward, but if you run into further issues feel free to [open a new issue](https://github.com/ndonfris/fish-lsp/issues)

No sorry method 2 worked. I forget to remove the lsp installed by mason at method 1 first. 

But anther question, I found that the treesitter cant be used. Mainly reason I need the fish-lsp is that I need to press `S` to get the function under cursor in a quick way.  With many empty line in a fish shell function make it very difficult to copy the function. 

What is your advice for this? 

---
update: oh it is the treesitter things. I fix it. thanks again

**Reply by ndonfris** [reply]
Glad you got it working!

My config is a little dated, but I like to use `s` for my surround action, and then I use [nvim-treesitter-textobjects](https://github.com/nvim-treesitter/nvim-treesitter-textobjects?tab=readme-ov-file#overriding-or-extending-textobjects) to handle selecting functions with `if`/`af` (inner function/around function).

That way I can do `yaf`/`vaf`/`dif`/`cif` to yank/select/delete/change the functions. 

Also, `%` is pretty good for jumping to closing or opening tokens.

---

Thanks for the detailed logs, those look to be insightful for why the mason build wasn't working. If I had to guess just going off of those, it looks like the build step of the lsp didn't fully succeed. 

By any chance, are you on one of the operating systems that is failing tests [here](https://github.com/mason-org/mason-registry/pull/8609)? Aside from the windows test, I'm not quite sure what caused those to fail.

**Reply by OriLiMu** [reply]
I use OS: Manjaro Linux x86_64
Kernel: Linux 6.6.80-1-MANJARO

Seems on the list

**Reply by ndonfris** [reply]
Thanks for the conformation :+1:

It does indeed look to be caused by the installation scripts using fish (which I assume isn't included in their testing containers).

I'll port the [postinstall scripts to node](https://github.com/ndonfris/fish-lsp/blob/d8780ab2fdc76af72a39106c3e81b11a2edfa215/package.json#L49) in the next release, and hopefully that will resolve the remaining blockers for installation to work out of the box inside mason.

**Comment by bnwa** [comment]
Haven't been able to attend to this and at any rate an official `fish-lsp` package is available from the registry now

**Reply by ndonfris** [reply]
Thank you for all the help :pray:
